File: app/src/main/java/ds/vkplus/ui/OnScrollBottomRecyclerViewListener.java
package ds.vkplus.ui;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import ds.vkplus.callback.IntCallback;
import ds.vkplus.utils.L;

public class OnScrollBottomRecyclerViewListener extends RecyclerView.OnScrollListener {

	private IntCallback cb;
	private LinearLayoutManager lm;

	public int prevLast = 0;
	private int speed = 0;
	//private int scrollState;
	private int last;
	private int total;


	public OnScrollBottomRecyclerViewListener(LinearLayoutManager lm, final IntCallback callback) {
		cb = callback;
		this.lm = lm;
	}


	private void checkResult() {
		if (last == total - 1 && prevLast != last) {
			L.v("onBottom");
			cb.onResult(last);
		}
	}


	/*public boolean isScrolling() {
		L.v("scroll state="+scrollState);
		return scrollState != SCROLL_STATE_IDLE;
	}*/


	@Override
	public void onScrollStateChanged(final RecyclerView recyclerView, final int newState) {
		//scrollState = newState;
		L.v("state=" + newState);
		if (newState == RecyclerView.SCROLL_STATE_IDLE) {
			//recyclerView.getAdapter().notifyItemRangeChanged(lm.findFirstVisibleItemPosition(), lm.findLastVisibleItemPosition());

			if (total < 3) {
				L.v("onBottom");
				cb.onResult(last);
			} /*else {
				((NewsFragment.NewsRecyclerAdapter) recyclerView.getAdapter()).resetInitTime();
				recyclerView.getAdapter().notifyItemRangeChanged(lm.findFirstVisibleItemPosition(), lm.findLastVisibleItemPosition());
			}*/
		}
	}


	@Override
	public void onScrolled(final RecyclerView recyclerView, final int dx, final int dy) {
		if (lm != null) {
			last = lm.findLastCompletelyVisibleItemPosition();
			if (last == -1)
				last = lm.findLastVisibleItemPosition();
			total = lm.getItemCount();
			//L.v("total=%s prevLast=%s last=%s sate=%s", total, prevLast, last, scrollState);
			checkResult();
			prevLast = last;
			this.speed = speed;
		}
	}


	public int getSpeed() {
		return speed;
	}

}

File: app/src/main/java/ds/vkplus/ui/OnScrollBottomRecyclerViewListener.kt
package ds.vkplus.ui

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import ds.vkplus.callback.IntCallback
import ds.vkplus.utils.L

public class OnScrollBottomRecyclerViewListener(private val lm: LinearLayoutManager, private val callback: (Int)->Unit) : RecyclerView.OnScrollListener() {
	
	public var prevLast: Int = 0
	public val speed: Int = 0
	//private int scrollState;
	private var last: Int = 0
	private var total: Int = 0
	
	
	private fun checkResult() {
		if (last == total - 1 && prevLast != last) {
			L.v("onBottom")
			callback(last)
		}
	}
	
	

	override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
		//scrollState = newState;
		L.v("state=" + newState)
		if (newState == RecyclerView.SCROLL_STATE_IDLE) {

			if (total < 3) {
				L.v("onBottom")
				callback(last)
			}
		}
	}
	
	
	override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
			last = lm.findLastCompletelyVisibleItemPosition()
			if (last == -1)
				last = lm.findLastVisibleItemPosition()
			total = lm.itemCount
			checkResult()
			prevLast = last
	}
	
}
